pkg/controllers/deprovisioning/controller.go

```diff
@@ -105,126 +108,64 @@ func (c *Controller) Builder(_ context.Context, m manager.Manager) controller.Bu
-		c.emptyNodeConsolidation,
-
-		// Attempt to identify multiple nodes that we can consolidate simultaneously to reduce pod churn
-		c.multiNodeConsolidation,
-
-		// And finally fall back our single node consolidation to further reduce cluster cost.
-		c.singleNodeConsolidation,
-	} {
+	// Attempt different deprovisioning methods. We'll only let one method perform an action
+	for _, d := range c.deprovisioners {
 		candidates, err := candidateNodes(ctx, c.cluster, c.kubeClient, c.clock, c.cloudProvider, d.ShouldDeprovision)
 		if err != nil {
-			return ResultFailed, fmt.Errorf("determining candidate nodes, %w", err)
+			return reconcile.Result{}, fmt.Errorf("determining candidate nodes, %w", err)
 		}
 		// If there are no candidate nodes, move to the next deprovisioner
 		if len(candidates) == 0 {
 			continue
 		}
 
-		result, err := c.executeDeprovisioning(ctx, d, candidates...)
+		// Determine the deprovisioning action
+		cmd, err := d.ComputeCommand(ctx, candidates...)
 		if err != nil {
-			return ResultFailed, fmt.Errorf("deprovisioning nodes, %w", err)
+			return reconcile.Result{}, fmt.Errorf("computing deprovisioning decision, %w", err)
 		}
-
-		switch result {
-		case ResultFailed:
-			return ResultFailed, err
-		case ResultRetry, ResultSuccess:
-			// the controller wants to retry, or was successful in deprovisioning
-			return result, nil
-		case ResultNothingToDo:
-			// found nothing to do, so try the next deprovisioner
+		if cmd.action == actionDoNothing {
 			continue
-		default:
-			logging.FromContext(ctx).Errorf("unexpected result %s", result)
 		}
+		if cmd.action == actionRetry {
+			return reconcile.Result{Requeue: true}, nil
+		}
+
+		// Attempt to deprovision
+		if err := c.executeCommand(ctx, d, cmd); err != nil {
+			return reconcile.Result{}, fmt.Errorf("deprovisioning nodes, %w", err)
+		}
+		return reconcile.Result{Requeue: true}, nil
 	}
 
+	// Mark cluster as consolidated if all deprovisioners succeeded
+	c.cluster.Consolidated()
+
 	// All deprovisioners did nothing, so return nothing to do
-	return ResultNothingToDo, nil
+	return reconcile.Result{RequeueAfter: pollingPeriod}, nil
 }
 
-// Given candidate nodes, compute best deprovisioning action
-func (c *Controller) executeDeprovisioning(ctx context.Context, d Deprovisioner, nodes ...CandidateNode) (Result, error) {
-	// Each attempt will try at least one node, limit to that many attempts.
-	cmd, err := d.ComputeCommand(ctx, nodes...)
-	if err != nil {
-		return ResultFailed, err
-	}
-	// Convert action to result
-	switch cmd.action {
-	case actionFailed:
-		return ResultFailed, err
-	case actionDoNothing:
-		return ResultNothingToDo, nil
-	case actionRetry:
-		return ResultRetry, nil
-	}
-	// If delete or replace, execute command
-	result, err := c.executeCommand(ctx, cmd, d)
-	if err != nil {
-		return ResultFailed, err
-	}
-	return result, nil
+// CandidateNode is a node that we are considering for deprovisioning along with extra information to be used in
```

Comment:
Can we consider moving `CandidateNode` to types?

Reply:
good idea!